File: 623-add-one-row/solution.js
/*
BFS看js 自己寫的
DFS看Go
 */
const TreeNode = require('../datastructure/Tree/TreeNode')
const buildLevelOrderTree_v2 = require('../datastructure/Tree/TreeDebugger_v2')
/**
 * @param {TreeNode} root
 * @param {number} val
 * @param {number} depth
 * @return {TreeNode}
 */
var addOneRow = function(root, val, depth) {
    if (depth === 1){
        let newNode = new TreeNode(val)
        newNode.left = root
        return newNode
    }
    let queue = [root]
    let curDepth = 0
    while (queue.length > 0){
        curDepth++
        let levelSize = queue.length
        for (let i = 0; i < levelSize; i++) {
            let [nextLeft, nextRight] = [null, null]
            let curNode = queue.shift()
            if (curDepth === depth - 1){
                //Step1
                if (curNode.left != null)
                    nextLeft = curNode.left
                if (curNode.right != null)
                    nextRight = curNode.right
                //Step2
                curNode.left = new TreeNode(val)
                curNode.right= new TreeNode(val)
                //Step3
                curNode.left.left = nextLeft
                curNode.right.right = nextRight
            }
            if (curNode.left != null)
                queue.push((curNode.left))
            if (curNode.right != null)
                queue.push(curNode.right)
        }
    }
    return root
};
if (require.main === module){
    const levelOrder = [4,2,6,3,1,5]
    const root = buildLevelOrderTree_v2(levelOrder)
    addOneRow(root, 1, 2)
}
/*
自己寫的
思路:
由上至下遍歷, 在 depth - 1 那層停下來, 執行插入

需要思考的是如何插入?
Step1. 先把 curNode.left, curNode.right 存起來, 避免等一下找不到
Step2. 再把 curNode.left, curNode.right 插入題目要求的節點
Step3. 再把step1. 存起來的那兩個節點賦值給 左新節點的左邊, 右新節點右邊

depth - 1 的每個節點都要做一次上述的動作, 就能達到題目要求
 */

File: 623-add-one-row/solution.go
/*
BFS看js 自己寫的
DFS看Go 自己寫的
 */
package main

import (
	"fmt"
	"gopractice/datastructure/Tree"
)
type TreeNode = Tree.TreeNode
func addOneRow(root *TreeNode, val int, depth int) *TreeNode {
	if depth == 1 {
		newNode := &TreeNode{Val: val}
		newNode.Left = root
		return newNode
	}
	var dfs func (*TreeNode, int) *TreeNode
	dfs = func(root *TreeNode, curDepth int) *TreeNode {
		if root == nil {
			return nil
		}
		if curDepth == depth - 1 {
			nextNode := root.Left
			newNode := &TreeNode{Val: val}
			root.Left = newNode
			newNode.Left = nextNode
		}
		dfs(root.Left, curDepth + 1)
		if curDepth == depth - 1 {
			nextNode := root.Right
			newNode := &TreeNode{Val: val}
			root.Right = newNode
			newNode.Right = nextNode
		}
		dfs(root.Right, curDepth + 1)
		return  nil
	}
	dfs(root, 1)
	return root
}
func main(){
	levelOrder := []any {4, 2, 6, 3, 1, 5}
	root := Tree.BuildLevelOrderTree_v2(levelOrder)
	fmt.Println("Ans = ", addOneRow(root, 1, 2))
}
/*
思路:
前序 + 中序插入新節點

用dfs遍歷整棵樹, 在碰到depth - 1 那層時把,目標節點的左邊給新增好,
因為是dfs所以還會再回到該目標節點一次, 此時就馬上把右邊也新增好, 如此就完成單一節點的新增, 
在遍歷的過程中每次回到 depth - 1 層, 就對當下節點做上述的動作, 循環往復就能達成題目的要求
*/